test/AutoDiff/method.swift

```diff
@@ -0,0 +1,305 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+
+var MethodTests = TestSuite("Method")
+
+// ==== Tests with generated adjoint ====
+
+struct Parameter : Equatable {
+  let x: Float
+}
+
+extension Parameter {
+  func squared() -> Float {
+    return x * x
+  }
+
+  static func squared(p: Parameter) -> Float {
+    return p.x * p.x
+  }
+
+  func multiplied(with other: Float) -> Float {
+    return x * other
+  }
+
+  static func * (_ a: Parameter, _ b: Parameter) -> Float {
+    return a.x * b.x
+  }
+}
+
+MethodTests.test("instance method with generated adjoint, called from differentated func") {
+  func f(_ p: Parameter) -> Float {
+    return 100 * p.squared()
+  }
+  let dF = #gradient(f)
+  expectEqual(Parameter(x: 4 * 100), dF(Parameter(x: 2)))
+  expectEqual(Parameter(x: 40 * 100), dF(Parameter(x: 20)))
+}
+
+MethodTests.test("instance method with generated adjoint, differentiated directly") {
+  // This is our current syntax for taking gradients of instance methods
+  // directly. If/when we develop nicer syntax for this, change this test.
+  let g = #gradient({ (p: Parameter) in p.squared() })
+  expectEqual(Parameter(x: 4), g(Parameter(x: 2)))
+  expectEqual(Parameter(x: 40), g(Parameter(x: 20)))
+}
+
+MethodTests.test("instance method with generated adjoint, wrt only self") {
+  func f(_ p: Parameter) -> Float {
+    return 100 * p.multiplied(with: 200)
+  }
+  let dF = #gradient(f)
+  expectEqual(Parameter(x: 100 * 200), dF(Parameter(x: 1)))
+  expectEqual(Parameter(x: 100 * 200), dF(Parameter(x: 2)))
+}
+
+MethodTests.test("instance method with generated adjoint, wrt only non-self") {
+  func f(_ other: Float) -> Float {
+    return 100 * Parameter(x: 200).multiplied(with: other)
+  }
+  let dF = #gradient(f)
+  expectEqual(100 * 200, dF(1))
+  expectEqual(100 * 200, dF(2))
+}
+
+MethodTests.test("instance method with generated adjoint, wrt self and non-self") {
+  let g = #gradient({ (p: Parameter, o: Float) in p.multiplied(with: o) })
+  expectEqual((Parameter(x: 100), 200), g(Parameter(x: 200), 100))
+  expectEqual((Parameter(x: 200), 100), g(Parameter(x: 100), 200))
+}
+
+MethodTests.test("static method with generated adjoint, called from differentiated func") {
+  func f(_ p: Parameter) -> Float {
+    return 100 * Parameter.squared(p: p)
+  }
+  let dF = #gradient(f)
+  expectEqual(Parameter(x: 4 * 100), dF(Parameter(x: 2)))
+  expectEqual(Parameter(x: 40 * 100), dF(Parameter(x: 20)))
+}
+
+// TODO(SR-8699): Fix this test.
```

Comment:
This test (and another below) is failing for reasons that seem unrelated to adjoint signature, so I will address their failures in a separate PR.